core/src/consensus.rs

```diff
@@ -347,6 +345,22 @@ impl Tower {
         last_vote
     }
 
+    fn maybe_timestamp(&mut self, current_slot: Slot) -> Option<UnixTimestamp> {
+        if current_slot > self.last_timestamp.slot {
+            let mut timestamp = Utc::now().timestamp();
+            if timestamp <= self.last_timestamp.timestamp {
+                timestamp = self.last_timestamp.timestamp + 1;
```

Comment:
Woot. we can't merge this... Under stable operation slots are 400ms, so we'll be always ahead of wall time...

Reply:
I'm not sure I follow.   At worse the timestamp could be a couple nanoseconds ahead of the wall time, but I don't see how that would matter

Reply:
@mvines Well, the granularity of `Utc::now().timestamp()` is _seconds_: https://docs.rs/chrono/0.4.11/chrono/struct.DateTime.html#method.timestamp. I guess, that's why this monotonic increasing guarantee code was added to begin with (CC: @CriesofCarrots )

Reply:
I don't grok this change, actually. I glanced at it earlier and thought it was in response to this logic (https://github.com/solana-labs/solana/blob/17a2128a8f7ebfb7ca8a22ebc3fd04b685975b59/programs/vote/src/vote_state/mod.rs#L547), but `process_timestamp` doesn't require that the timestamp advance. It just requires that either the slot or the timestamp advance.

Here's an example of what I think @ryoqun was pointing out:
- Say Vote for slot 1 at T0 has `Utc::now().timestamp() = 1593450853`
- Vote for slot 2 at T0 + 400ms also has `Utc::now().timestamp() = 1593450853`, therefore this logic updates the timestamp to 1593450854
- Vote for slot 3 at T0 + 800ms also has `Utc::now().timestamp() = 1593450853`, therefore this logic updates the timestamp to 1593450855 (stored last_timestamp + 1). The stored timestamp is now 2sec ahead, and if the validator continues voting on time, the stored timestamp will continue gaining 2sec/real sec.

Reply:
> Well, the granularity of Utc::now().timestamp() is seconds:

😱 omg, ok then that's broken.  I incorrectly understood it to be `ns`

Reply:
Heh, good time for in-source comments? :)

Reply:
I reworked `maybe_timestamp()` with the latest commit.   It now only timestamps votes at most once a second, which is still more than plenty to avoid the concern that spawned this issue